Use new index interface review round 1

postamar opened the pull request:
this is here just for CI

Commit:
sql: add Index interface and TableDescriptor methods in catalog

Previously, the catalog.TableDescriptor interface and its implementing
types would liberally return descpb.IndexDescriptor values, pointers and
slices. In an effort to stop manipulating such protos directly, this
patch introduces a catalog.Index interface to encapsulate it. In order
to enventually propagate this change throughout the code base, this
patch marks some existing catalog.TableDescriptor methods as deprecated
and introduces new ones to eventually replace them.

Partially addresses #57465.

Release note: None

Commit:
sql: remove deprecated (Writable|Deletable|DeleteOnly)Indexes methods

Previously, these methods would be called on a table descriptor interface
(or backing struct) to obtain a slice of descpb.IndexDescriptors. This
patch removes these calls, along with the method definitions, in favour
of new methods which use the catalog.Index interface type instead.

Partially addresses #57465.

Release note: None

Commit:
sql: remove deprecated AllNonDropIndexes method

Previously, this method would be called on a table descriptor interface
(or backing struct) to obtain a slice of descpb.IndexDescriptor pointers.
This patch removes these calls, along with the method definitions, in
favour of new methods which use the catalog.Index interface type instead.

Partially addresses #57465.

Release note: None

Commit:
sql: remove deprecated ForeachNonDropIndex method

Previously, this method would be called on a table descriptor interface
(or backing struct) to apply a function on the *descpb.IndexDescriptor
of each of the table's non-drop indexes.
This patch removes these calls, along with the method definition, in
favour of new methods which use the catalog.Index interface type instead.

Partially addresses #57465.

Release note: None

Commit:
sql: remove deprecated ForeachIndex method

Previously, this method would be called on a table descriptor interface
(or backing struct) to apply a function on the *descpb.IndexDescriptor
of each of the table's indexes.
This patch removes these calls, along with the method definition, in
favour of new methods which use the catalog.Index interface type instead.

Partially addresses #57465.

Release note: None

Commit:
sql: remove deprecated FindIndexByID method

Previously, this method would be called on a table descriptor interface
(or backing struct) to return a target *descpb.IndexDescriptor.
This patch removes these calls, along with the method definition, in
favour of new methods which use the catalog.Index interface type instead.

Partially addresses #57465.

Release note: None

Commit:
sql: remove deprecated FindIndexByName method

Previously, this method would be called on a table descriptor interface
(or backing struct) to return a target *descpb.IndexDescriptor.
This patch removes these calls, along with the method definition, in
favour of new methods which use the catalog.Index interface type instead.

Partially addresses #57465.

Release note: None

Commit:
sql: remove deprecated GetPublicNonPrimaryIndexes method

Previously, this method would be called on a table descriptor interface
(or backing struct) to obtain a slice of descpb.IndexDescriptors. This
patch removes these calls, along with the method definition, in favour
of new methods which use the catalog.Index interface type instead.

Partially addresses #57465.

Release note: None

Commit:
sql: remove deprecated GetPrimaryIndex method

Previously, this method would be called on a table descriptor interface
(or backing struct) to obtain the primary index descriptor in the form
of a *descpb.IndexDescriptor.
This patch removes these calls, along with the method definition, in
favour of new methods which use the catalog.Index interface type instead.

Partially addresses #57465.

Release note: None

Commit:
sql: rename PrimaryIndexInterface method to GetPrimaryIndex

Previously, the name PrimaryIndexInterface had been chosen because
a deprecated GetPrimaryIndex() method already existed and because
PrimaryIndex is an existing field in descpb.IndexDescriptor which is
embedded in the structs implementing the catalog.TableDescriptor
interface.

This patch renames the method to GetPrimaryIndex: although its return
type differs it has the same meaning as the recently-removed method of
the same name.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/58503)
<!-- Reviewable:end -->

Commit:
sql: improve index caching in tabledesc.Immutable

This patch changes the indexCache struct in a bid to reduce the number
of memory allocations that take place when accessing its contents. The
previous behaviour of GetPrimaryIndex() was particularly bad in that it
resulted in an allocation in every call, which is easily avoided.

Release note: None

Commit:
sql: turn some catalog.TableDescriptor methods into functions

The catalog.TableDescriptor has methods which can just as well be
expressed as functions, thereby lightening that interface definition
somewhat.

This patch also adds comments to many catalog.Index methods.

Release note: None

Commit:
sql: add unit testing to new catalog.Index interface

This commit introduces unit tests for the new catalog.Index interface,
which were missing in my earlier commits.

Release note: None
